### ISSUE TITLE
 Does Loguru have built-in support for running code before/after logging a message?

### ISSUE THREAD

**Issue by dougthor42** [Delgan/loguru#375]
Does Loguru have built-in support for running code before/after logging a message? What I'm trying to do is get `loguru.logger.warning` to also emit `warnings.warn` when called.

The use case is that I have a lot of calls to `loguru.logger.warning` throughout the project. I've now been asked to update things so that the exit code is non-zero if a warning is emitted. For example:

```python
def cli():
    exit_code = 0  # success
    with warnings.capture_warnings(record=True) as w:
        try:
            main.run()
        except Exception:
            exit_code = 1  # error

        if len(w) > 0:
            exit_code = 3  # success-with-warnings
    sys.exit(exit_code)
```

What I could do is go find all instances of `logger.warning(msg)` and add a line `warnings.warn(msg)` afterwards, but it seems like there's gotta be a better way.

If there's nothing built-in, I'll end up doing something like:

```python
# -*- coding: utf-8 -*-
# __init__.py
import warnings

# Normally I'd just have from `loguru import logger` here
from loguru._logger import Logger as _Logger
from loguru._logger import Core as _Core

# Monkeypatch the loguru logger so that logger.warning also emits warning.warn
class CustomLogger(_Logger):
    def warning(__self, __message, *args, **kwargs):
        r"""Log ``message.format(*args, **kwargs)`` with severity ``'WARNING'``."""
        __self._log("WARNING", None, False, __self._options, __message, args, kwargs)
        warnings.warn(__message)

# basically the same as loguru/__init__.py
logger = CustomLogger(
    core=_Core(),
    exception=None,
    depth=0,
    record=False,
    lazy=False,
    colors=False,
    raw=False,
    capture=True,
    patcher=None,
    extra={},
)
```

**Comment by Delgan** [comment]
Could you simply use `warnings.warn` as a basic sink?

```python
import warnings

from loguru import logger

logger.add(warnings.warn, format="{message}", filter=lambda record: record["level"] == "WARNING")
```

**Comment by dougthor42** [comment]
Ah, I knew there would be an easier way! That does sound like a much simpler and more elegant solution.

That said, I tested it and doesn't seem to work. Let me investigate more and try and figure out the issue.

**Comment by dougthor42** [comment]
Here's a short example showing that adding the sink with `warnings.warn` doesn't work. The sink is correctly made, but it looks like the `warnings.warn` function isn't being called.

```python
# loguru_warning.py
import warnings
from loguru import logger

logger.add(warnings.warn, format="{message}", filter=lambda record: record["level"] == "WARNING")

# prove that our sink is configured:
logger.info(logger._core.handlers)

# two warnings should be recorded:
with warnings.catch_warnings(record=True) as captured_warnings:
    logger.warning("foo")
    warnings.warn("bar")
    for w in captured_warnings:
        logger.info(f"   Captured warning: {w}")
    assert len(captured_warnings) == 2
```

```
$ python loguru_warning.py
2020-12-08 14:55:40.138 | INFO     | __main__:<module>:7 - {0: (id=0, level=10, sink=<stderr>), 1: (id=1, level=10, sink=warn)}
2020-12-08 14:55:40.139 | WARNING  | __main__:<module>:11 - foo
2020-12-08 14:55:40.139 | INFO     | __main__:<module>:14 -    Captured warning: {message : UserWarning('bar',), category : 'UserWarning', filename : 'loguru_warning.py', lineno : 12, line : None}
Traceback (most recent call last):
  File "loguru_warning.py", line 15, in <module>
    assert len(captured_warnings) == 2
AssertionError
```

I tried replacing:
```diff
-logger.add(warnings.warn, format="{message}", filter=lambda record: record["level"] == "WARNING")
+logger.add(lambda msg: print(msg), format="{message}", filter=lambda record: record["level"] == "WARNING")
```
and the "foo" was still not printed. So I'm inclined to think that Loguru isn't correctly calling the function, but I'm sure it's just that I'm missing something else.

### Version details:
+ Python 3.6.8 x64 on Ubuntu 18.04 on WSL
+ Loguru 0.5.0

**Comment by Delgan** [comment]
Sorry @dougthor42, it was a typo in the snippet I shared! I changed the code a few minutes after posting the message, but apparently it was too late.

In the `filter` function, you need to specifically compare the `name` of the level: 

```python
lambda record: record["level"].name == "WARNING"`
```

**Comment by dougthor42** [comment]
Ah, yes, that indeed did the trick! I guess I should have been able to figure that out by looking at the docs, haha.

So this resolves my particular usage case. Do you want to keep this Issue open to discuss support for what essentially is pre- and post-logging hooks? Or do you think that custom sinks indirectly satisfy that need? Personally, I'm thinking that they do and this issue can be closed.

**Comment by Delgan** [comment]
Yeah, I think sinks make a good solution for this purpose. Unless there are use cases for which they are not enough, I prefer not to add a new method to the `logger`. Maybe it's a bit indirect and tricky, but I think I'll just add this snippet in the documentation. :+1:

**Comment by Delgan** [comment]
I'm closing this issue now that I added the `warnings.warn()` example to the documentation. :+1:
